bundles/org.dataflowanalysis.analysis.converter/src/org/dataflowanalysis/analysis/converter/PalladioConverter.java

```diff
@@ -0,0 +1,184 @@
+            Node previousNode = null;
+            for (AbstractActionSequenceElement<?> ASE : actionSequence.getElements()) {
+                if (ASE instanceof AbstractPCMActionSequenceElement) {
+                    previousNode = processActionSequenceElement((AbstractPCMActionSequenceElement<?>) ASE, previousNode);
+                }
+            }
+        }
+        return new DataFlowDiagramAndDictionary(dataFlowDiagram, dataDictionary);
+    }
+
+    private Node processActionSequenceElement(AbstractPCMActionSequenceElement<? extends Entity> pcmASE, Node previousDFDNode) {
+        Node dfdNode = getOrCreateDFDNode(pcmASE);
+
+        createFlowBetweenPreviousAndCurrentNode(previousDFDNode, dfdNode, pcmASE);
+
+        return dfdNode;
+    }
+
+    private void createFlowBetweenPreviousAndCurrentNode(Node source, Node dest, AbstractPCMActionSequenceElement<? extends Entity> pcmASE) {
+        if (source == null || dest == null) {
+            return;
+        }
+        List<DataFlowVariable> FlowVariables = pcmASE.getAllDataFlowVariables();
+        for (DataFlowVariable flowVariable : FlowVariables) {
+            String flowName = flowVariable.variableName();
+
+            Optional<Flow> optFlow = dataFlowDiagram.getFlows().stream().filter(f -> f.getSourceNode().equals(source))
+                    .filter(f -> f.getDestinationNode().equals(dest)).filter(f -> f.getEntityName().equals(flowName)).findFirst();
+
+            if (optFlow.isPresent()) {
+                return;
+            }
+
+            Flow newFlow = dataflowdiagramFactory.eINSTANCE.createFlow();
+            newFlow.setSourceNode(source);
+            newFlow.setDestinationNode(dest);
+            newFlow.setEntityName(flowName);
+
+            // Palladio Assumption: Each flows between two nodes with the same parameters/that are called the same use the same pin
+            Pin sourceOutPin = findOrCreateOutputPin(source, flowName);
+            Pin destInPin = findOrCreateInputPin(dest, flowName);
+            newFlow.setSourcePin(sourceOutPin);
+            newFlow.setDestinationPin(destInPin);
+
+            ForwardingAssignment forwarding = datadictionaryFactory.eINSTANCE.createForwardingAssignment();
+            forwarding.setOutputPin(sourceOutPin);
+            source.getBehaviour().getAssignment().add(forwarding);
+
+            this.dataFlowDiagram.getFlows().add(newFlow);
+        }
+    }
+
+    // A pin is equivalent if the same parameters are passed
+    private Pin findOrCreateOutputPin(Node source, String parameters) {
+        Optional<Pin> optPin = source.getBehaviour().getOutPin().stream().filter(p -> p.getEntityName().equals(parameters)).findAny();
+        if (optPin.isPresent()) {
+            return optPin.get();
+        }
+        Pin pin = datadictionaryFactory.eINSTANCE.createPin();
+        pin.setEntityName(parameters);
+        source.getBehaviour().getOutPin().add(pin);
+        return pin;
```

Comment:
Personally, I would replace the present check with the following:
Return the opt pin result, but call `orElse` on the Optional and extract the case of creating a pin into a method (You can also pass a reference to an In/Out Pin to unify the behavior with the one in the following method)